/mob/living/proc/Life(delta_time)

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Makes `/mob/living/proc/Life` and most related procs use `delta_time`
- Procs that had snowflaked timing systems, such as breathing, addiction, and advanced diseases were left unchanged.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- More consistent code

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
refactor: Mobcode is now ever so slightly more consistent
/:cl:

- [x] TODO: Testing
  - Some testing done. Didn't notice any major PR-related runtimes and everything appeared to be behaving as it should.
- [x] TODO: figure out what's failing the unit tests
  - Turns out the metabolization and stomach unit tests both assumed that reagents would metabolize at a rate of `/datum/reagent/var/metabolization_rate` per tick. Given that the metabolization rate now factors in `delta_time` the mob was metabolizing all of the reagents before the unit test could check if they were in the mob. The unit tests have been updated to factor in SSmobs' wait.

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Swapping most of Life() over to use delta_time

- Makes SSmobs calculate delta_time and pass it along to life
- Converts most Life() related procs to use delta_time

Jared-Fogle commented on code/__DEFINES/atmospherics.dm:
```diff
@@ -135,15 +135,15 @@
 /// Similar to the BODYTEMP_AUTORECOVERY_DIVISOR, but this is the divisor which is applied at the stage that follows autorecovery. This is the divisor which comes into play when the human's loc temperature is higher than their body temperature. Make it lower to gain bodytemp faster.
 #define BODYTEMP_HEAT_DIVISOR				15
-/// The maximum number of degrees that your body can cool in 1 tick, due to the environment, when in a cold area.
+/// The maximum number of degrees that your body can cool in 1 tixk, due to the environment, when in a cold area.
```
wuh

TemporalOroboros replied:
I think I changed it to seconds at some point, and then changed it back and missed.

TemporalOroboros replied:
Fixed

TemporalOroboros committed:
Fixes tixks

- Corrects a comment I changed back and missed

TemporalOroboros committed:
Fixes unit tests

- It turns out the reagents metabolism and stomach unit tests both assumed that reagents would always metabolize `/datum/reagent/var/metabolization_rate` per tick. And changing the system to use `delta_time` meant the liver was now metabolizing all of the allocated reagents before the unit tests got around to checking whether the reagents were in the mob.

Donkie commented:
Very nice 👍

Ghommie commented on code/__DEFINES/mobs.dm:
```diff
@@ -75,13 +75,13 @@
-#define HEAT_DAMAGE_LEVEL_2 3 //Amount of damage applied when your body temperature passes the 400K point
-#define HEAT_DAMAGE_LEVEL_3 8 //Amount of damage applied when your body temperature passes the 460K point and you are on fire
+#define HEAT_DAMAGE_LEVEL_1 1 //Amount of damage applied when your body temperature just passes the 360.15k safety point
+#define HEAT_DAMAGE_LEVEL_2 2.5 //Amount of damage applied when your body temperature passes the 400K point
```
```suggestion
#define HEAT_DAMAGE_LEVEL_2 1.5 //Amount of damage applied when your body temperature passes the 400K point
```

Donkie commented on code/datums/diseases/pierrot_throat.dm:
```diff
@@ -4,31 +4,31 @@
 	cure_text = "Banana products, especially banana bread."
 	cures = list(/datum/reagent/consumable/banana)
-	cure_chance = 75
+	cure_chance = 30
```
Fix probability

Donkie commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 		if(1)
 			if(restcure)
-				if(affected_mob.body_position == LYING_DOWN && prob(30))
+				if(affected_mob.body_position == LYING_DOWN && DT_PROB(15, delta_time))
```
Fix probability

Donkie commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 				to_chat(affected_mob, "<span class='danger'>Your entire body vibrates.</span>")
 
-			if (prob(35))
+			if (DT_PROB(20, delta_time))
```
Fix probability

Donkie commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 					cure()
 					return FALSE
-			if (prob(60))
+			if (DT_PROB(35, delta_time))
```
Fix probability

Donkie commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -10,33 +10,33 @@
 	switch(stage)
 		if(2)
-			if(prob(45))
+			if(DT_PROB(25, delta_time))
```
Fix probability

Donkie commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -10,33 +10,33 @@
 		if(5)
 			to_chat(affected_mob, "<span class='danger'>Your body is unable to contain the Rhumba Beat...</span>")
-			if(prob(50))
+			if(DT_PROB(30, delta_time))
```
Fix probability

Donkie commented on code/datums/diseases/transformation.dm:
```diff
@@ -219,26 +219,26 @@
 	cure_text = "frost oil"
 	cures = list(/datum/reagent/consumable/frostoil)
-	cure_chance = 80
+	cure_chance = 40
```
Fix probability

Donkie commented on code/datums/diseases/transformation.dm:
```diff
@@ -319,8 +319,8 @@
 	cures = list(/datum/reagent/consumable/condensedcapsaicin) //beats the hippie crap right out of your system
-	cure_chance = 80
-	stage_prob = 5
+	cure_chance = 40
```
Fix probability

Donkie commented on code/datums/mutations/actions.dm:
```diff
@@ -167,10 +167,10 @@
 	if(!isturf(owner.loc))
 		return
-	if(prob((0.5+((100-dna.stability)/20))) * GET_MUTATION_SYNCHRONIZER(src)) //very rare, but enough to annoy you hopefully. +0.5 probability for every 10 points lost in stability
+	if(DT_PROB((0.25+((100-dna.stability)/40)) * GET_MUTATION_SYNCHRONIZER(src), delta_time)) //very rare, but enough to annoy you hopefully. +0.5 probability for every 10 points lost in stability
```
Fix probability, might be abit more difficult here but perhaps tune the 0.25 and /40 to better adapt to common dna.stability values

TemporalOroboros replied:
The issue here is that common dna stability values are 0-100

TemporalOroboros replied:
I did the math and the maximum difference between the two is about 0.03%
I'm not going to bother with a more complicated calculation.

Donkie commented on code/modules/antagonists/revenant/revenant_blight.dm:
```diff
@@ -1,12 +1,12 @@
 	spread_text = "A burst of unholy energy"
 	cures = list(/datum/reagent/water/holywater)
-	cure_chance = 50 //higher chance to cure, because revenants are assholes
+	cure_chance = 25 //higher chance to cure, because revenants are assholes
```
Fix probability

Donkie commented on code/modules/antagonists/revenant/revenant_blight.dm:
```diff
@@ -25,42 +25,42 @@
+			affected_mob.adjustToxLoss(1 * stage * delta_time, FALSE) //should, normally, do about 30 toxin damage.
 			new /obj/effect/temp_visual/revenant(affected_mob.loc)
-		if(prob(45))
+		if(DT_PROB(25, delta_time))
```
Fix probability

Donkie commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1165,19 +1165,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 		else if(H.satiety < 0)
 			H.satiety++
-			if(prob(round(-H.satiety/40)))
+			if(DT_PROB(round(-H.satiety/80), delta_time))
```
Fix probability, might be difficult here again

TemporalOroboros replied:
Corrected. The maximum deviation is now ~0.07%

Donkie reviewed (commented):
Added comments about probabilities that needs fixing. Might have missed some, and might have commented on too many, but keep a lookout for these things!

TemporalOroboros committed:
Fixes heat damage coefficients

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

TemporalOroboros committed:
Fixes a bunch of probabilies to be more accurate, also honey.

- Corrects a bunch of probabilites that drifted when I changed them over to DT_PROB. The values have been rounded to the nearest integer if over 10.
- Fixes honey double dipping delta_time

TemporalOroboros commented on code/datums/diseases/_disease.dm:
```diff
@@ -15,14 +15,17 @@
-	var/infectivity = 65
-	var/cure_chance = 8
+	/// The probability of spreading through the air every second
+	var/infectivity = 30
```
Done,

TemporalOroboros committed:
Merge branch 'life-delta-time' of https://github.com/TemporalOroboros/tgstation into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros commented:
Have done a bit of testing. Didn't notice any major PR-related runtimes and everything seemed to be behaving as it should. I think this one's ready if anyone wants it.

TemporalOroboros committed:
The great REM-ining

- Applies REM to every consistent reagent effect in on_mob_life and overdose_process
- This does not include prob() calls. Those are still just DT_PROB

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Fix missing delta_time

- Re-adds an arg that got removed when merging master

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Unpoison granabitaluri

- Fix inverted healing

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into upstream/master

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Revert "Merge branch 'master' into life-delta-time"

This reverts commit ded378a15d2f8262fe6885d784b126202fa06a6c.

TemporalOroboros committed:
Revert "Revert "Merge branch 'master' into life-delta-time""

This reverts commit 03645cab9da84ae40492f3917b0ca271433fb519.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Re-add monkies because I accidentally deleted them resolving merge conflicts.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge remote-tracking branch 'origin/life-delta-time' into life-delta-time

TemporalOroboros committed:
Fixes space indentation

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Fixes master merge issues

TemporalOroboros committed:
Refix bodytemperature cold damage

TemporalOroboros committed:
Reimplements damage type and mod for cold temps

LemonInTheDark commented on code/__DEFINES/misc.dm:
```diff
@@ -465,7 +465,12 @@ GLOBAL_LIST_INIT(pda_styles, sortList(list(MONO, VT, ORBITRON, SHARE)))
 
 #define STANDARD_GRAVITY 1 //Anything above this is high gravity, anything below no grav
-#define GRAVITY_DAMAGE_TRESHOLD 3 //Starting with this value gravity will start to damage mobs
+/// The gravity strength threshold for high gravity damage.
+#define GRAVITY_DAMAGE_THRESHOLD 3
+/// The scaling factor for high gravity damage.
+#define GRAVITY_DAMAGE_SCALING	0.5
+/// The maximum [BRUTE] damage a mob can take from high gravity per second.
+#define GRAVITY_DAMAGE_MAXIMUM	1.5
```
Oh dear god not another gravity rework

TemporalOroboros replied:
This is just defining the things that were already there.

LemonInTheDark commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -32,11 +32,12 @@ SUBSYSTEM_DEF(mobs)
 	//cache for sanic speed (lists are references anyways)
 	var/list/currentrun = src.currentrun
 	var/times_fired = src.times_fired
+	var/delta_time = wait / (1 SECONDS)	// TODO: Make this actually responsive to stuff like pausing and resuming
```
If you plan on making delta time reflect subsystems taking too long to fire you're going to have a hell of a time. Am I misunderstanding you?

TemporalOroboros replied:
Not at all. That's exactly what I had in mind.

TemporalOroboros replied:
It's not a current project though.

LemonInTheDark commented on code/datums/brain_damage/magic.dm:
```diff
@@ -95,7 +102,7 @@
 		owner.visible_message("<span class='warning'>[owner] is torn apart by invisible claws!</span>", "<span class='userdanger'>Ghostly claws tear your body apart!</span>")
 		owner.take_bodypart_damage(rand(20, 45), wound_bonus=CANT_WOUND)
-	else if(prob(50))
+	else if(DT_PROB(30, delta_time))
```
Shouldn't this be 25?

TemporalOroboros replied:
The more accurate probabilities aren't exactly half of the original probabilities. I mostly just used desmos to figure them out.

Donkie replied:
To clarify, you solve for x to find the new rate:

1-(1-x)^OLDDT=OLDCHANCE

So in this case:

1-(1-x)^2=0.5
=> x=~0.3

LemonInTheDark replied:
good lord

LemonInTheDark commented on code/datums/brain_damage/phobia.dm:
```diff
@@ -31,16 +34,17 @@
-		next_check = world.time + 50
-		var/list/seen_atoms = view(7, owner)
 
+	if(COOLDOWN_FINISHED(src, check_cooldown) && COOLDOWN_FINISHED(src, scare_cooldown))
```
You mind early returning here?

TemporalOroboros replied:
Not at all.

LemonInTheDark commented on code/datums/brain_damage/special.dm:
```diff
@@ -53,21 +53,25 @@
+	COOLDOWN_DECLARE(portal_cooldown)
+
+/datum/brain_trauma/special/bluespace_prophet/on_life(delta_time, times_fired)
+	if(COOLDOWN_FINISHED(src, portal_cooldown))
```
Also seems like a good candidate

TemporalOroboros replied:
Done

LemonInTheDark commented on code/datums/brain_damage/special.dm:
```diff
@@ -129,17 +133,18 @@
+/datum/brain_trauma/special/quantum_alignment/on_life(delta_time, times_fired)
 	if(linked)
 		if(QDELETED(linked_target))
 			linked_target = null
 			linked = FALSE
-		else if(!returning && world.time > snapback_time)
+		else if(!returning && COOLDOWN_FINISHED(src, snapback_cooldown))
 			start_snapback()
 		return
-	if(prob(4))
+	if(DT_PROB(2, delta_time))
 		try_entangle()
```
Ditto? kinda

TemporalOroboros replied:
Agreed

LemonInTheDark commented on code/datums/brain_damage/split_personality.dm:
```diff
@@ -131,7 +131,7 @@
 		trauma = _trauma
 	return ..()
 
-/mob/living/split_personality/Life()
```
Why do this?

TemporalOroboros replied:
So that it conforms with the rest,

LemonInTheDark commented on code/datums/diseases/_disease.dm:
```diff
@@ -15,14 +15,17 @@
-	var/infectivity = 65
-	var/cure_chance = 8
+	/// The probability of spreading through the air every second
+	var/infectivity = 41
```
Why did 65 become 41?

TemporalOroboros replied:
Because it's more accurate than 32.5

LemonInTheDark commented on code/datums/diseases/advance/advance.dm:
```diff
@@ -254,8 +254,8 @@
 		permeability_mod = max(CEILING(0.4 * properties["transmittable"], 1), 1)
-		cure_chance = 15 - clamp(properties["resistance"], -5, 5) // can be between 10 and 20
-		stage_prob = max(properties["stage_rate"], 2)
+		cure_chance = clamp(7.5 - (0.5 * properties["resistance"]), 5, 10) // can be between 5 and 10
+		stage_prob = max(0.5 * properties["stage_rate"], 1)
```
What the hell is going on here?

TemporalOroboros replied:
Making these probabilities also conform with the proper rates.

LemonInTheDark commented on code/datums/diseases/advance/symptoms/heal.dm:
```diff
@@ -160,9 +160,7 @@
-	C.reagents.metabolize(C, can_overdose=TRUE) //this works even without a liver; it's intentional since the virus is metabolizing by itself
-	if(triple_metabolism)
-		C.reagents.metabolize(C, can_overdose=TRUE)
+	C.reagents.metabolize(C, triple_metabolism ? 4 : 2, 0, can_overdose=TRUE) //this works even without a liver; it's intentional since the virus is metabolizing by itself
```
I do not like this

TemporalOroboros replied:
Looking at it again I don't either.

TemporalOroboros replied:
It's just about the best I can do without passing through delta_time and symptoms have their own time handling system.

LemonInTheDark commented on code/datums/diseases/anxiety.dm:
```diff
@@ -12,32 +12,32 @@
+			if(DT_PROB(2.5, delta_time))
 				to_chat(affected_mob, "<span class='notice'>You feel panicky.</span>")
-			if(prob(2))
+			if(DT_PROB(2, delta_time))
```
Shouldn't this be a 1?

Donkie replied:
Yes

TemporalOroboros replied:
Yes

LemonInTheDark commented on code/datums/diseases/cold.dm:
```diff
@@ -10,47 +10,47 @@
 	switch(stage)
 		if(2)
-			if(affected_mob.body_position == LYING_DOWN && prob(40))  //changed FROM prob(10) until sleeping is fixed
+			if(affected_mob.body_position == LYING_DOWN && DT_PROB(23, delta_time))  //changed FROM prob(10) until sleeping is fixed // Has sleeping been fixed yet?
```
Shouldn't this be 20?

TemporalOroboros replied:
Nope

LemonInTheDark commented on code/datums/diseases/decloning.dm:
```diff
@@ -24,38 +24,38 @@
-			if(prob(2))
+			if(DT_PROB(1, delta_time))
 				affected_mob.emote("drool")
 			if(prob(3))
```
Missed one?

TemporalOroboros replied:
It appears so

TemporalOroboros replied:
Fixed

LemonInTheDark commented on code/datums/diseases/flu.dm:
```diff
@@ -4,52 +4,52 @@
-			if(prob(1))
+			if(DT_PROB(0.5, delta_time))
 				to_chat(affected_mob, "<span class='danger'>Your muscles ache.</span>")
 				if(prob(20))
```
Missed one? Maybe?

Donkie replied:
No, the outer prob already fixes the time dependency, so using DT_PROB here would cause wrong behaviour.

TemporalOroboros replied:
The outer one handles it. Applying it twice would halve it.

LemonInTheDark commented on code/datums/diseases/gbs.dm:
```diff
@@ -5,29 +5,29 @@
 		if(4)
 			to_chat(affected_mob, "<span class='userdanger'>Your body feels as if it's trying to rip itself apart!</span>")
-			if(prob(50))
+			if(DT_PROB(30, delta_time))
```
30 != 50 / 2

LemonInTheDark commented on code/datums/diseases/parasitic_infection.dm:
```diff
@@ -28,20 +28,20 @@
 				affected_mob.adjust_nutrition(-6)
 		if(4)
-			if(prob(30))
+			if(DT_PROB(16, delta_time))
```
16 != 30 / 2

LemonInTheDark commented on code/datums/diseases/parrotpossession.dm:
```diff
@@ -25,7 +25,7 @@
 		return FALSE
 
-	if(length(parrot.speech_buffer) && prob(parrot.speak_chance))
+	if(length(parrot.speech_buffer) && DT_PROB(parrot.speak_chance, delta_time))	// I'm not going to dive into polycode trying to adjust that probability. Enjoy doubled ghost parrot speach
```
Didn't you need to do this anyway? Also can't you just do the * 0.5 trick?

TemporalOroboros replied:
Multiplying by 0.5 only works with small probabilities, and even then it's only mostly accurate. Also, speak chance is handled by the npc subsystem and I'm already overhauling SSmobs time handling. I don't want to do SSnpcpool too.

LemonInTheDark commented on code/datums/diseases/pierrot_throat.dm:
```diff
@@ -4,31 +4,31 @@
 	cure_text = "Banana products, especially banana bread."
 	cures = list(/datum/reagent/consumable/banana)
-	cure_chance = 75
+	cure_chance = 50
```
50 != 75 / 2

LemonInTheDark commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 		if(1)
 			if(restcure)
-				if(affected_mob.body_position == LYING_DOWN && prob(30))
+				if(affected_mob.body_position == LYING_DOWN && DT_PROB(16, delta_time))
```
16 != 30 / 2

LemonInTheDark commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 				to_chat(affected_mob, "<span class='danger'>Your entire body vibrates.</span>")
 
-			if (prob(35))
+			if (DT_PROB(19, delta_time))
```
19 != 35 / 2

LemonInTheDark commented on code/datums/diseases/retrovirus.dm:
```diff
@@ -26,61 +26,61 @@
 					cure()
 					return FALSE
-			if (prob(60))
+			if (DT_PROB(37, delta_time))
```
37 != 60 / 2

LemonInTheDark commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -10,33 +10,33 @@
 	switch(stage)
 		if(2)
-			if(prob(45))
+			if(DT_PROB(26, delta_time))
```
26 != 45 / 2

TemporalOroboros replied:
It's more accurate.

LemonInTheDark commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -10,33 +10,33 @@
-			if(prob(20))
-				if (prob(50))
+			if(DT_PROB(10, delta_time))
+				if(prob(50))
```
Missed one?

TemporalOroboros replied:
It's inside of the `DT_PROB` if I also changed it I'd halve the effective rate.

LemonInTheDark commented on code/datums/diseases/rhumba_beat.dm:
```diff
@@ -10,33 +10,33 @@
 		if(5)
 			to_chat(affected_mob, "<span class='danger'>Your body is unable to contain the Rhumba Beat...</span>")
-			if(prob(50))
+			if(DT_PROB(29, delta_time))
```
29 != 50 / 2

TemporalOroboros replied:
More accurate.

LemonInTheDark commented on code/datums/diseases/transformation.dm:
```diff
@@ -218,26 +218,26 @@
 	cure_text = "frost oil"
 	cures = list(/datum/reagent/consumable/frostoil)
-	cure_chance = 80
+	cure_chance = 55
```
55 != 80 / 2

TemporalOroboros replied:
Again, it's more accurate.

LemonInTheDark commented on code/datums/diseases/transformation.dm:
```diff
@@ -318,8 +318,8 @@
 	cures = list(/datum/reagent/consumable/condensedcapsaicin) //beats the hippie crap right out of your system
-	cure_chance = 80
-	stage_prob = 5
+	cure_chance = 55
```
Ditto

TemporalOroboros replied:
Ditto

LemonInTheDark reviewed (changes requested):
I have reviewed 47 / 188 files, or up to the end of disease code. More coming when I have slept.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

Donkie commented on code/modules/unit_tests/stomach.dm:
```diff
@@ -17,9 +17,9 @@
 
 	//Give them meth and let it kick in
-	pill.reagents.add_reagent(meth, initial(meth.metabolization_rate) * 1.9)
+	pill.reagents.add_reagent(meth, 1.9 * initial(meth.metabolization_rate) * initial(SSmobs.wait) / (1 SECONDS))
```
Use SSMOBS_DT instead

https://github.com/tgstation/tgstation/blob/d0a4840669eb4c4fa65deba9b5927d48d65bb492/code/__DEFINES/subsystems.dm#L251

TemporalOroboros replied:
Done

Donkie commented on code/modules/unit_tests/metabolizing.dm:
```diff
@@ -22,15 +22,15 @@
 
 	// Give them enough meth to be consumed in 2 metabolizations
-	pill.reagents.add_reagent(meth, initial(meth.metabolization_rate) * 1.9)
+	pill.reagents.add_reagent(meth, 1.9 * initial(meth.metabolization_rate) * initial(SSmobs.wait) / (1 SECONDS))
```
Use SSMOBS_DT instead

https://github.com/tgstation/tgstation/blob/d0a4840669eb4c4fa65deba9b5927d48d65bb492/code/__DEFINES/subsystems.dm#L251

TemporalOroboros replied:
Done

TemporalOroboros committed:
Fixes a few probs I missed.

Donkie commented on code/modules/surgery/organs/stomach.dm:
```diff
@@ -68,9 +68,9 @@
 
 		// Transfer the amount of reagents based on volume with a min amount of 1u
-		var/amount = min(round(metabolism_efficiency * bit.volume, 0.1) + amount_min, amount_max)
+		var/amount = min((round(metabolism_efficiency * amount_max, 0.05) + rate_min) * delta_time, amount_max)
```
Pretty confused about the logic here, why is a rate added onto the computed amount?

Also, why have you compensated all the metabolization_efficiency vars and not the rates? Feels like rates would change with DT but efficiencies should stay constant.

TemporalOroboros replied:
Because what is confusingly called efficiency is actually how quickly the stomach transferrs the reagents to the mobs bloodstream. The max rate is a limiter to make sure we don't try to take more reagent out than there exists.

Donkie replied:
Alright, should probably be renamed abit then, but not a job for this PR

TemporalOroboros committed:
SSMOBS_DT

Cyberboss commented:
Yeeting this onto Bagil

Cyberboss commented:
Adding the rest of the fleet

Cyberboss commented:
> 
> 
> I have reviewed 47 / 188 files, or up to the end of disease code. More coming when I have slept.

Makes me wish GitHub had better reviewing tools so we could each take a chunk.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

antropod commented:
I think thermal regulation is broken with this PR.
It was test merged on Manuel and people were freezing in normal conditions. One of examples round ID 154520, and a lot people froze to death.
Room is 22 degrees, people are dying because their core temperature is -5.2 degrees. Other people have their body and core temperature at 65 degrees and its not stabilizing.

Jared-Fogle commented:
What @antropod said--we've been getting a lot of reports of people dying due to cold extremely fast/when they don't think they should have been, specifically on servers with this TM'd.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

Jared-Fogle commented:
Okay, that may not actually be the fault of this PR and just a happy coincidence--heard some Terry players complaining about it more when the TM wasn't even in 🤷

TemporalOroboros committed:
Cleans up some bodytemp procs

LemonInTheDark commented:
It might be a mix of cold spreading more aggressively and bodytemp issues, I've seen some cases where people are dying to heat in a much cooler room, and their bodytemp is behaving oddly, then again it might just be the slow cooling thing bodytemp does.

upsetpanini1337 commented:
had an issue recently where I was put through the brainwashing machine that sec gets twice (I managed to slide out the second time before it was complete) and my entire vision was swaying radically further than it should've been. I was hot, hungry, at low sanity and also brainwashed at the time, so... pls fix

TemporalOroboros commented:
I honestly have zero idea what could cause the bodytemp issues.

LemonInTheDark commented:
We've entered the endgame

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

antropod commented:
I ran old and new versions with debugger and this is results:
https://github.com/tgstation/tgstation/blob/77a8af8b8f73f31e662f62009e301a53444d1af1/code/modules/mob/living/carbon/human/species.dm#L1705
https://github.com/tgstation/tgstation/blob/77a8af8b8f73f31e662f62009e301a53444d1af1/code/modules/mob/living/carbon/human/species.dm#L1724
these two lines use function `get_temp_change_amount` which uses `BODYTEMP_AUTORECOVERY_DIVISOR` which you changed from 28 to 14

This somehow breaks equlibrium and causes body temperature to stabilize at different values.

For example, with old code your body temperature was stabilizing at 31C at normal conditions and now it stays at 37C.
Also with new code core and body temperatures are always same, with old code they were different.

TemporalOroboros commented:
I graphed out the bodytemp handling and got the same thing. I'll fix it.

TemporalOroboros committed:
Should fix erratic bodytemps

- People should no longer spontaneously freeze or broil to death.
- Contains miscellaneous tidyness changes.

TemporalOroboros commented:
Right, actually need to push the fix.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

RaveRadbury commented:
This is causing Blood Deficiency to never tick below 99%, it just hovers between 100% and 99%

TemporalOroboros committed:
Fixes doubled blood regen

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Makes blob mobs conform to delta_time again

Cyberboss reviewed (changes requested):
Fix conflicts for testmerge

LemonInTheDark commented on code/datums/mutations/body.dm:
```diff
@@ -9,8 +9,8 @@
-/datum/mutation/human/epilepsy/on_life()
-	if(prob(1 * GET_MUTATION_SYNCHRONIZER(src)) && owner.stat == CONSCIOUS)
+/datum/mutation/human/epilepsy/on_life(delta_time, times_fired)
+	if(DT_PROB(0.5 * GET_MUTATION_SYNCHRONIZER(src), delta_time) && owner.stat == CONSCIOUS)
```
I'm just gonna go ahead and assume all of these sorts of changes are sane, otherwise I'm gonna go insane.  Is that a reasonable assumption to make?

TemporalOroboros replied:
Yes. For both.

LemonInTheDark commented on code/datums/mutations/body.dm:
```diff
@@ -336,15 +336,16 @@
+	if(DT_PROB(13, delta_time))
+		if(COOLDOWN_FINISHED(src, msgcooldown))
 			to_chat(owner, "<span class='danger'>Your acid flesh bubbles...</span>")
-			msgcooldown = world.time + 200
-		if(prob(15))
-			owner.acid_act(rand(30,50), 10)
+			COOLDOWN_START(src, msgcooldown, 20 SECONDS)
+		if(DT_PROB(7.5, delta_time))
```
Forgive me if my memory is going bust, but I recall something about nested probs like this only needing the first one converted, is that incorrect?

TemporalOroboros replied:
You are correct. I'll fix that.

LemonInTheDark commented on code/datums/wounds/burns.dm:
```diff
@@ -70,42 +70,41 @@
-			else if(disabling && prob(8))
+			else if(disabling && DT_PROB(4, delta_time))
 				to_chat(victim, "<span class='notice'>You regain sensation in your [limb.name], but it's still in terrible shape!</span>")
 				set_disabling(FALSE)
-			else if(prob(20))
+			else if(DT_PROB(10, delta_time))
 				victim.adjustToxLoss(0.5)
 		if(WOUND_INFECTION_CRITICAL to WOUND_INFECTION_SEPTIC)
-			if(!disabling && prob(3))
+			if(!disabling && DT_PROB(1.5, delta_time))
 				to_chat(victim, "<span class='warning'><b>You suddenly lose all sensation of the festering infection in your [limb.name]!</b></span>")
 				set_disabling(TRUE)
-			else if(disabling && prob(3))
+			else if(disabling && DT_PROB(1.5, delta_time))
 				to_chat(victim, "<span class='notice'>You can barely feel your [limb.name] again, and you have to strain to retain motor control!</span>")
 				set_disabling(FALSE)
-			else if(prob(1))
+			else if(DT_PROB(0.5, delta_time))
 				to_chat(victim, "<span class='warning'>You contemplate life without your [limb.name]...</span>")
 				victim.adjustToxLoss(0.75)
-			else if(prob(4))
+			else if(DT_PROB(2, delta_time))
 				victim.adjustToxLoss(1)
```
Is this actually the proper conversion for else ifs? Could be wrong, but if wait is raised the probabilities wouldn't stay the same, would they?

TemporalOroboros replied:
You're right. I'll fix it.

LemonInTheDark commented on code/datums/wounds/pierce.dm:
```diff
@@ -89,13 +89,13 @@
-/datum/wound/pierce/on_xadone(power)
+/datum/wound/pierce/on_xadone(power, delta_time, times_fired)
 	. = ..()
-	blood_flow -= 0.03 * power // i think it's like a minimum of 3 power, so .09 blood_flow reduction per tick is pretty good for 0 effort
+	blood_flow -= 0.03 * power * delta_time // i think it's like a minimum of 3 power, so .09 blood_flow reduction per tick is pretty good for 0 effort
 
 /datum/wound/pierce/on_synthflesh(power)
 	. = ..()
-	blood_flow -= 0.05 * power // 20u * 0.05 = -1 blood flow, less than with slashes but still good considering smaller bleed rates
+	blood_flow -= 0.025 * power // 20u * 0.05 = -1 blood flow, less than with slashes but still good considering smaller bleed rates
```
Why do these two deal with the delta time conversion differently? The result of how they're used?

TemporalOroboros replied:
I think I just forgot the delta_time for one of them.

TemporalOroboros replied:
Done. I moved the delta time out of the x-adone proc.

LemonInTheDark commented on code/datums/wounds/slash.dm:
```diff
@@ -103,15 +103,15 @@
 
 	if(HAS_TRAIT(victim, TRAIT_BLOODY_MESS))
-		blood_flow += 0.5 // old heparin used to just add +2 bleed stacks per tick, this adds 0.5 bleed flow to all open cuts which is probably even stronger as long as you can cut them first
+		blood_flow += 0.25 // old herapin used to just add +2 bleed stacks per tick, this adds 0.5 bleed flow to all open cuts which is probably even stronger as long as you can cut them first
```
Is this proper?

TemporalOroboros replied:
Let me check. I last touched this almost two weeks ago.

TemporalOroboros replied:
Fixed

LemonInTheDark reviewed (changes requested):
Mostly questions about the logic behind some changes, I've thrown in the towel on checking the exact values of probs, just gonna keep at this till it's merged or I reach the end. Read up to the end of wounds

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Fixes master merge, random code cleanup, and some of Lemon's suggestions.

- Fixes bananium golem cooldowns
- Makes temperature damage respect delta_time
- Makes x-adone wound effects use delta_time
- Makes species metabolism overrides use delta_time
- Herapin -> Heparin
- Fixes some wound code scaling

TemporalOroboros commented:
I did some testing. It works.

LemonInTheDark reviewed (changes requested):
This is more stuff adjacent to your code. tho there's one timing issue I think? Honestly not sure. 80/193

TemporalOroboros commented on code/modules/antagonists/blob/blobstrains/regenerative_materia.dm:
```diff
@@ -24,8 +24,8 @@
-/datum/reagent/blob/regenerative_materia/on_mob_life(mob/living/carbon/C)
-	C.adjustToxLoss(1*REAGENTS_EFFECT_MULTIPLIER)
+/datum/reagent/blob/regenerative_materia/on_mob_life(mob/living/carbon/C, delta_time, times_fired)
+	C.adjustToxLoss(1 * REAGENTS_EFFECT_MULTIPLIER * delta_time)
```
So do I. But I think it reads more clearly if its explicit rather than implied.

TemporalOroboros commented on code/modules/antagonists/revenant/revenant_blight.dm:
```diff
@@ -25,42 +25,42 @@
-		if(affected_mob.body_position == LYING_DOWN && prob(stage * 6))
+		if(affected_mob.body_position == LYING_DOWN && DT_PROB(3 * stage, delta_time))
 			cure()
 			return FALSE
-		if(prob(stage * 3))
+		if(DT_PROB(1.5 * stage, delta_time))
 			to_chat(affected_mob, "<span class='revennotice'>You suddenly feel [pick("sick and tired", "disoriented", "tired and confused", "nauseated", "faint", "dizzy")]...</span>")
 			affected_mob.add_confusion(8)
 			affected_mob.adjustStaminaLoss(20, FALSE)
 			new /obj/effect/temp_visual/revenant(affected_mob.loc)
 		if(stagedamage < stage)
 			stagedamage++
-			affected_mob.adjustToxLoss(stage * 2, FALSE) //should, normally, do about 30 toxin damage.
+			affected_mob.adjustToxLoss(1 * stage * delta_time, FALSE) //should, normally, do about 30 toxin damage.
```
If you turn up the delta_time it will end up becoming more and more likely to just delete itself on the first tick and not actually do anything. The only way I can see to fix that would be to move that bit until after it handles the effects, which would be it's own behavior change.

TemporalOroboros replied:
Done

TemporalOroboros commented on code/modules/antagonists/blob/blob_mobs.dm:
```diff
@@ -278,39 +278,42 @@
+				H.color = "#000000"
+
+	if(damagesources)
+		for(var/i in 1 to damagesources)
+			adjustHealth(maxHealth*BLOBMOB_BLOBBERNAUT_HEALTH_DECAY * delta_time) //take 2.5% of max health as damage when not near the blob or if the naut has no factory, 5% if both
```
It probably just sends signals. Maybe damage messages. I'll look into it.

TemporalOroboros replied:
Done

TemporalOroboros commented on code/datums/mutations/radioactive.dm:
```diff
@@ -9,8 +9,8 @@
-/datum/mutation/human/radioactive/on_life()
-	radiation_pulse(owner, 20 * GET_MUTATION_POWER(src))
+/datum/mutation/human/radioactive/on_life(delta_time, times_fired)
+	radiation_pulse(owner, 10 * GET_MUTATION_POWER(src) * delta_time)
```
I think that's basically just contamination? I'm not all too familiar with the effects of radiation and a quick skim doesn't show anything that would be a major issue. I guess in theory if something managed to hold SSmobs up for long enough this could generate a pulse large enough to hang SSmobs up for longer which would cause an even stronger pulse etc, etc. But, that's for when I get around to making SSmobs scale with lag.

TemporalOroboros replied:
Looking at radiation code I don't think this will cause any major issues. I'll fix it if it does.

LemonInTheDark replied:
Anturk's right, the strength of the radiation pulse is directly effects how far it can travel, and how much it can contaminate.
Converting it over like this will cause issues later if the wait of Life is ever raised, or if you implement that dynamic scaling idea of yours. It might just be best to leave it alone honestly.

TemporalOroboros replied:
That's mostly a failure of how radiation works, but I see what you mean. I'll look into it.

LemonInTheDark replied:
Have you addressed this?

TemporalOroboros replied:
Even if dynamic delta time was implemented it would take SSmobs locking up for over 35 seconds for this to start contaminating things. I really don't think it will be a problem. Or at least not without much more pressing issues already present.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Reorders disease operations. Also blobbernaut damage scaling is a multiplier instead of a loop.

- Some diseases had a chance to cure themselves at random during normal processing. Hypothetically, if someone turned the delta_time for SSmobs extremely high this would make these diseases almost invariably cure themselves before they actually did anything. This has been changed so they process symptoms before the chance to cure themselves.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

Cyberboss commented:
fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

Qustinnus commented:
there is no conflicts

Cyberboss commented:
You can't see them in draft mode.

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros commented:
@Cyberboss This should be ready to TM again after it passes CI.

Cyberboss commented:
Mark it ready next time please

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

Cyberboss commented:
Conflicts are not your friend

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros commented:
No they are not.

LemonInTheDark commented on code/modules/mob/living/carbon/alien/larva/life.dm:
```diff
@@ -1,13 +1,13 @@
-		if(amount_grown < max_grown)
-			amount_grown++
-			update_icons()
+	if(!..() || IS_IN_STASIS(src) || (amount_grown >= max_grown)) //not dead and not in stasis
```
Update the comment?

TemporalOroboros replied:
Sure

TemporalOroboros replied:
Done.

LemonInTheDark commented on code/modules/mob/living/carbon/alien/larva/life.dm:
```diff
@@ -1,13 +1,13 @@
+	if(!..() || IS_IN_STASIS(src) || (amount_grown >= max_grown)) //not dead and not in stasis
+		return
+	// GROW!
+	amount_grown = min(amount_grown + (1 * delta_time), max_grown)
```
Doesn't this double growth speed?

TemporalOroboros replied:
Probably. I'll fix it.

LemonInTheDark commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -18,34 +18,33 @@
-/mob/living/carbon/human/Life()
-	if (notransform)
+/mob/living/carbon/human/Life(delta_time = SSMOBS_DT, times_fired)
+	if(notransform)
```
I don't really care myself, but for the future avoid changing these when you find them. It's purely a matter of personal taste and at a large enough scale it'll be you changing them one way, and someone else changing them back in 2 years. Better to just let it sit

TemporalOroboros replied:
I mean, yes. But it's slightly compulsory at this point.

LemonInTheDark replied:
Actually it's pure neutral, the guidelines straight up say not to touch it.

LemonInTheDark commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -303,14 +302,17 @@
-				lastpuke = 0
+	if(stat)
+		return
+	if(getToxLoss() < 45 || nutrition < 20)
```
```suggestion
	if(getToxLoss() < 45 || nutrition <= 20)
```

TemporalOroboros replied:
Done.

LemonInTheDark commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -303,14 +302,17 @@
+	if(getToxLoss() < 45 || nutrition < 20)
+		return
+
+	lastpuke += DT_PROB(30, delta_time)
+	if(lastpuke >= 50) // about 25 second delay I guess // This is actually closer to 150 seconds
+		vomit(20)
+		lastpuke = 0
```
What the fuck

TemporalOroboros replied:
Also my reaction.

TemporalOroboros committed:
Whitespace fixes

LemonInTheDark commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1190,19 +1190,19 @@ GLOBAL_LIST_EMPTY(roundstart_races)
-			H.overeatduration++
+		if(H.overeatduration < 2 MINUTES) //capped so people don't take forever to unfat
+			H.overeatduration += delta_time
 	else
-		if(H.overeatduration > 1)
-			H.overeatduration -= 2 //doubled the unfat rate
+		if(H.overeatduration > 0)
+			H.overeatduration -= 2 * delta_time //doubled the unfat rate
```
What's going on here?

TemporalOroboros replied:
I thought that, since it was a timer, it should be in time units. I've given it another pass.

TemporalOroboros committed:
Some of potato's suggestions

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -1150,16 +1153,16 @@
 	description = "Drastically increases movement speed, but deals toxin damage."
 	color = "#AE151D"
-	metabolization_rate = 1
+	metabolization_rate = 2.5 * REAGENTS_METABOLISM
```
?

TemporalOroboros replied:
I'm using REAGENTS_METABOLISM to scale metabolization rates.

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -1240,49 +1243,49 @@
+		M.adjustStaminaLoss(-0.5 * REM * delta_time, 0)
 		M.Jitter(1)
-		metabolization_rate = 0.01 * REAGENTS_METABOLISM * rand(5,20) // randomizes metabolism between 0.02 and 0.08 per tick
+		metabolization_rate = 0.005 * REAGENTS_METABOLISM * rand(5, 20) // randomizes metabolism between 0.02 and 0.08 per tick
```
Per second?

LemonInTheDark replied:
Do you mind updating this comment?

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/drink_reagents.dm:
```diff
@@ -64,15 +64,17 @@
 			//nothing
-		if(21 to INFINITY)
-			if(prob(current_cycle-10))
+		if(21 to 110)
+			if(DT_PROB(100 * (1 - (sqrt(110 - current_cycle) / 10)), delta_time))
 				M.cure_nearsighted(list(EYE_DAMAGE))
```
What the fuck is going on here

TemporalOroboros replied:
Math

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/drink_reagents.dm:
```diff
@@ -333,15 +335,15 @@
 	color = "#FFB766"
 	quality = DRINK_NICE
-	nutriment_factor = 2
+	nutriment_factor = 2 * REAGENTS_METABOLISM
```
?

TemporalOroboros replied:
Scaling factor

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/drink_reagents.dm:
```diff
@@ -398,13 +400,13 @@
-	M.adjust_bodytemperature(-5 * TEMPERATURE_DAMAGE_COEFFICIENT, M.get_body_temp_normal())
+	M.adjust_bodytemperature(-5 * REM * TEMPERATURE_DAMAGE_COEFFICIENT * delta_time, M.get_body_temp_normal())
 	..()
 	. = 1
```
You mind setting this to TRUE? I assume that's what it's meant to be

TemporalOroboros replied:
No problem

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -197,36 +195,36 @@
-	M.adjustToxLoss(1, 0)
-	M.adjustOrganLoss(ORGAN_SLOT_BRAIN, pick(0.5, 0.6, 0.7, 0.8, 0.9, 1))
+	M.adjustToxLoss(1 * REM * delta_time, 0)
+	M.adjustOrganLoss(ORGAN_SLOT_BRAIN, (rand(5, 10) / 100) * REM * delta_time)
```
Shouldn't you divide by 10 here?

TemporalOroboros replied:
Yep

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2,7 +2,7 @@
 	name = "Blood"
 	color = "#C80000" // rgb: 200, 0, 0
-	metabolization_rate = 5 //fast rate so it disappears fast.
+	metabolization_rate = 12.5 * REAGENTS_METABOLISM //fast rate so it disappears fast.
```
What?

TemporalOroboros replied:
Adding in a scaling factor

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -335,23 +335,23 @@
+		M.adjustToxLoss(-1*delta_time, 0)
+		M.adjustOxyLoss(-1*delta_time, 0)
+		M.adjustBruteLoss(-1*delta_time, 0)
+		M.adjustFireLoss(-1*delta_time, 0)
 		if(ishuman(M) && M.blood_volume < BLOOD_VOLUME_NORMAL)
-			M.blood_volume += 3
+			M.blood_volume += 1.5*delta_time
 	else  // Will deal about 90 damage when 50 units are thrown
-		M.adjustOrganLoss(ORGAN_SLOT_BRAIN, 3, 150)
-		M.adjustToxLoss(2, 0)
-		M.adjustFireLoss(2, 0)
-		M.adjustOxyLoss(2, 0)
-		M.adjustBruteLoss(2, 0)
+		M.adjustOrganLoss(ORGAN_SLOT_BRAIN, 1.5*delta_time, 150)
+		M.adjustToxLoss(1*delta_time, 0)
+		M.adjustFireLoss(1*delta_time, 0)
+		M.adjustOxyLoss(1*delta_time, 0)
+		M.adjustBruteLoss(1*delta_time, 0)
```
Why not use REM here?

TemporalOroboros replied:
Because I learned about rem afterwards.

TemporalOroboros replied:
Fixed

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/toxin_reagents.dm:
```diff
@@ -422,17 +422,17 @@
-			M.Sleeping(40)
-			M.adjustToxLoss((current_cycle - 50)*REM, 0)
+			M.Sleeping(40 * REM * delta_time)
+			M.adjustToxLoss(1 * (current_cycle - 50) * REM * delta_time, 0)
```
Is the 1 for readability? You don't do similar things in similar situations

TemporalOroboros replied:
Pretty much.

LemonInTheDark commented on code/modules/reagents/chemistry/reagents/toxin_reagents.dm:
```diff
@@ -993,17 +993,17 @@
 	reagent_state = LIQUID
 	metabolization_rate = 0 //stays in the system until active.
-	var/actual_metaboliztion_rate = REAGENTS_METABOLISM
+	var/actual_metaboliztion_rate = 0.5 * REAGENTS_METABOLISM
```
?

TemporalOroboros replied:
This was done before I messed with REAGENTS_METABOLISM

LemonInTheDark reviewed (changes requested):
194/194
I'm going to blame mothblocks for this, even though it's my own fault.

To be totally honest with you, I'm not sure delta_time was a good idea in the first place, since it complicates things somewhat, but maybe that's just my smool brain.

In any case, if we're implementing something to make conversions later painless, it should do what it says on the tin.

TemporalOroboros commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -59,7 +56,7 @@
 			next_breath--
 
 	if((times_fired % next_breath) == 0 || failed_last_breath)
-		breathe() //Breathe per 4 ticks if healthy, down to 2 if our lungs or heart are damaged, unless suffocating
+		breathe(delta_time, times_fired) //Breathe per 4 ticks if healthy, down to 2 if our lungs or heart are damaged, unless suffocating
 		if(failed_last_breath)
 			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "suffocation", /datum/mood_event/suffocation)
```
I would like to say yes, but man breathcode needs a whole PR of its own.

TemporalOroboros commented on code/modules/mob/living/life.dm:
```diff
@@ -33,75 +41,76 @@
+	if(temp_delta < 0) // it is cold here
 		if(!on_fire) // do not reduce body temp when on fire
-			adjust_bodytemperature(max((loc_temp - bodytemperature) / BODYTEMP_DIVISOR, BODYTEMP_COOLING_MAX))
+			adjust_bodytemperature(max(max(temp_delta / BODYTEMP_DIVISOR, BODYTEMP_COOLING_MAX) * delta_time, temp_delta))
```
One, because we already have the temperature delta and we might as well use it. Two, I thought I would protect against overshooting in case someone slows SSmobs way down.

TemporalOroboros commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1721,7 +1735,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 		// Get the changes to the skin from the core temp
 		var/core_skin_diff = humi.coretemperature - humi.bodytemperature
 		// change rate of 0.09 to reflect temp back to the skin at the slight higher rate then core to skin
```
Done.

TemporalOroboros commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1702,7 +1716,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	var/area_skin_diff = area_temp - humi.bodytemperature
 	if(!humi.on_fire || area_skin_diff > 0)
 		// change rate of 0.1 as area temp has large impact on the surface
```
Done

TemporalOroboros commented on code/modules/mob/living/silicon/robot/life.dm:
```diff
@@ -1,24 +1,26 @@
 		if(cell.charge <= 100)
 			uneq_all()
-		var/amt = clamp((lamp_enabled * lamp_intensity),1,cell.charge) //Lamp will use a max of 5 charge, depending on brightness of lamp. If lamp is off, borg systems consume 1 point of charge, or the rest of the cell if it's lower than that.
+		var/amt = clamp(lamp_enabled * lamp_intensity * delta_time, 1, cell.charge) //Lamp will use a max of 5 charge, depending on brightness of lamp. If lamp is off, borg systems consume 1 point of charge, or the rest of the cell if it's lower than that.
```
Yes.

TemporalOroboros commented on code/modules/mob/living/simple_animal/friendly/cat.dm:
```diff
@@ -224,7 +224,7 @@
 			for(var/mob/living/simple_animal/mouse/M in view(1,src))
 				if(istype(M, /mob/living/simple_animal/mouse/brown/tom) && inept_hunter)
-					if (emote_cooldown < (world.time - 600))
+					if(emote_cooldown < (world.time - 600))
```
Sure.

TemporalOroboros commented on code/modules/mob/living/simple_animal/simple_animal.dm:
```diff
@@ -385,44 +385,52 @@
-			var/diff = areatemp - bodytemperature
-			diff = diff / 5
-			adjust_bodytemperature(diff)
+		var/temp_delta = areatemp - bodytemperature
+		if(abs(temp_delta) > 5)
+			if(temp_delta < 0)
+				if(!on_fire)
+					adjust_bodytemperature(clamp(temp_delta * delta_time / 10, temp_delta, 0))
+			else
+				adjust_bodytemperature(clamp(temp_delta * delta_time / 10, 0, temp_delta))
```
Yes.

TemporalOroboros commented on code/modules/mob/living/simple_animal/simple_animal.dm:
```diff
@@ -385,44 +385,52 @@
+		. = TRUE
+
+	if((bodytemperature > maxbodytemp) && unsuitable_heat_damage)
+		adjustHealth(unsuitable_heat_damage * delta_time)
 		switch(unsuitable_heat_damage)
 			if(1 to 5)
 				throw_alert("temp", /atom/movable/screen/alert/hot, 1)
 			if(5 to 10)
 				throw_alert("temp", /atom/movable/screen/alert/hot, 2)
 			if(10 to INFINITY)
 				throw_alert("temp", /atom/movable/screen/alert/hot, 3)
-	else
+		. = TRUE
+
+	if(!.)
```
Because I didn't see any harm in making it possible for mobs to be both overheating and freezing at the same time.

TemporalOroboros commented on code/modules/mob/living/simple_animal/slime/life.dm:
```diff
@@ -220,27 +226,27 @@
 		return
 
-	add_nutrition((rand(7, 15) * CONFIG_GET(number/damage_multiplier)))
+	add_nutrition((rand(3.5 * delta_time, 7.5 * delta_time) * CONFIG_GET(number/damage_multiplier)))
```
Point

TemporalOroboros replied:
Done

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -221,16 +221,16 @@ All effects don't start immediately, but rather get worse over time; the rate is
 	description = "A potent mixture of caffeine and alcohol."
 	color = "#102000" // rgb: 16, 32, 0
-	nutriment_factor = 1 * REAGENTS_METABOLISM
+	nutriment_factor = 2 * REAGENTS_METABOLISM
```
I think I did that before I touched the define and forgot to switch it back.

TemporalOroboros commented on code/modules/unit_tests/reagent_mod_expose.dm:
```diff
@@ -25,7 +25,7 @@
 	drink.attack(human, human)
 	TEST_ASSERT_EQUAL(human.fire_stacks, 1, "Human does not have fire stacks after taking phlogiston")
-	human.Life()
+	human.Life((2 SECONDS) / (1 SECONDS))
```
Yep

TemporalOroboros commented on code/modules/unit_tests/pills.dm:
```diff
@@ -5,6 +5,6 @@
 
 	pill.attack(human, human)
-	human.Life()
+	human.Life((2 SECONDS) / (1 SECONDS))
```
Yep

TemporalOroboros commented on code/modules/unit_tests/metabolizing.dm:
```diff
@@ -9,7 +9,7 @@
 /datum/unit_test/metabolization/proc/test_reagent(mob/living/carbon/C, reagent_type)
 	C.reagents.add_reagent(reagent_type, 10)
-	C.reagents.metabolize(C, can_overdose = TRUE)
+	C.reagents.metabolize(C, 2, 0, can_overdose = TRUE)
```
Yep

TemporalOroboros commented on code/modules/surgery/organs/heart.dm:
```diff
@@ -245,12 +245,14 @@
 	name = "heart of freedom"
 	desc = "This heart pumps with the passion to give... something freedom."
 	organ_flags = ORGAN_SYNTHETIC //the power of freedom prevents heart attacks
+	///
+	COOLDOWN_DECLARE(adrenaline_cooldown)
 	var/min_next_adrenaline = 0
```
I just forgot to remove it.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -254,28 +255,28 @@
+			heating = 20
 			if(isslime(M))
-				heating = rand(20,25)
-	M.adjust_bodytemperature(heating)
+				heating = rand(20, 25)
+	M.adjust_bodytemperature(heating * TEMPERATURE_DAMAGE_COEFFICIENT * REM * delta_time)
```
Yeah.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/cat2_medicine_reagents.dm:
```diff
@@ -111,14 +111,14 @@
-/datum/reagent/medicine/c2/probital/overdose_process(mob/living/M)
-	M.adjustStaminaLoss(3*REM, 0)
+/datum/reagent/medicine/c2/probital/overdose_process(mob/living/M, delta_time, times_fired)
+	M.adjustStaminaLoss(6 * REM * delta_time, 0)
```
I think I doubled it before I did the pass where I added REM to everything.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/cat2_medicine_reagents.dm:
```diff
@@ -184,17 +184,17 @@
-	M.reagents?.chem_temp +=(-10*REM)
-	M.adjust_fire_stacks(-1)
+		humi.adjust_coretemperature(rand(-25,-5) * TEMPERATURE_DAMAGE_COEFFICIENT * REM * delta_time, 50)
+	M.reagents?.chem_temp += (-10 * REM * delta_time)
+	M.adjust_fire_stacks(-1 * REM * delta_time)
```
REM is basically an auto-scaler for reagent effects. It was basically designed to serve the same purpose as delta_time. I just halved it and tacked along with delta_time onto everything without it. Everything that already had it got delta_time.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/cat2_medicine_reagents.dm:
```diff
@@ -208,11 +208,11 @@
 	if(ishuman(M))
 		var/mob/living/carbon/human/humi = M
-		humi.adjust_coretemperature(-10 * TEMPERATURE_DAMAGE_COEFFICIENT*REM,50)
+		humi.adjust_coretemperature(-10 * TEMPERATURE_DAMAGE_COEFFICIENT * REM * delta_time, 50)
```
The min and max are how cold/hot it can make the mob. I don't think we want those to change with SSmobs

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/cat2_medicine_reagents.dm:
```diff
@@ -282,32 +283,31 @@
 		if(chemtemp < radbonustemp*0.1) //if you're super chilly, it takes off 25% of your current rads
-			M.radiation = round(M.radiation * 0.75)
+			M.radiation = round(M.radiation * (0.75**(REM * delta_time)))
 		else if(chemtemp < radbonustemp)//else if you're under the chill-zone, it takes off 10% of your current rads
-			M.radiation = round(M.radiation * 0.9)
+			M.radiation = round(M.radiation * (0.90**(REM * delta_time)))
 		M.radiation -= radcalc
```
I am trying to understand what I was thinking here.
I think this is another case of "did it before I made everything REM"

TemporalOroboros replied:
Oh wait, no. That's right. Because exponential scaling.

LemonInTheDark replied:
What exponential scaling?

TemporalOroboros replied:
Because the effect is multiplicative rather than additive/subtractive the proper way to scale it is by exponentiating rather than multiplying.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -698,23 +701,23 @@
 		to_chat(M, "<span class='warning'>The blackness in your peripheral vision fades.</span>")
 		M.cure_nearsighted(EYE_DAMAGE)
-		M.blur_eyes(10)
+		M.blur_eyes(5)
```
It has been two months since I touched this part of the code. Let me check.

TemporalOroboros replied:
It looks like I didn't bother. I think the logic here is that, since it's basically a one-time event, it doesn't need to be delta-timed. Like, it's just blurring your eyes when it cures them and it's not exactly doing that over time.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -943,21 +946,21 @@
-	M.AdjustAllImmobility(-60)
-	M.adjustStaminaLoss(-5*REM, 0)
+		M.adjustOxyLoss(-1 * REM * delta_time, 0)
+		M.adjustToxLoss(-1 * REM * delta_time, 0)
+		M.adjustBruteLoss(-1 * REM * delta_time, 0)
+		M.adjustFireLoss(-1 * REM * delta_time, 0)
+	M.AdjustAllImmobility(-40  * REM * delta_time)
```
Yeah

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -971,10 +974,10 @@
+	if(M.AdjustSleeping(-20 * REM * delta_time))
 		. = 1
-	holder.remove_reagent(/datum/reagent/consumable/sugar, 3)
+	holder.remove_reagent(/datum/reagent/consumable/sugar, 1.5*delta_time)
```
Yes

TemporalOroboros commented on code/modules/mob/living/living.dm:
```diff
@@ -1857,7 +1857,7 @@
 
 	// Trait removal if obese
 	if(HAS_TRAIT_FROM(src, TRAIT_FAT, OBESITY))
-		if(overeatduration >= 100)
```
I did!

TemporalOroboros commented on code/modules/mob/living/silicon/robot/life.dm:
```diff
@@ -1,24 +1,26 @@
 			uneq_all()
-		var/amt = clamp((lamp_enabled * lamp_intensity),1,cell.charge) //Lamp will use a max of 5 charge, depending on brightness of lamp. If lamp is off, borg systems consume 1 point of charge, or the rest of the cell if it's lower than that.
+		var/amt = clamp(lamp_enabled * lamp_intensity * delta_time, 1, cell.charge) //Lamp will use a max of 5 charge, depending on brightness of lamp. If lamp is off, borg systems consume 1 point of charge, or the rest of the cell if it's lower than that.
 		cell.use(amt) //Usage table: 1/tick if off/lowest setting, 4 = 4/tick, 6 = 8/tick, 8 = 12/tick, 10 = 16/tick
```
Done

TemporalOroboros committed:
More of potato's requests and properly making overeatduration use SECONDS

Cyberboss commented:
@LemonInTheDark if this looks good to you, slap the TM label on it please

Cyberboss commented:
Oh god conflicts again 😢

Mothblocks commented:
>  I'm going to blame mothblocks for this, even though it's my own fault.

WHAT

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Applies delta_time to master merges

TemporalOroboros committed:
Readds reagent processing delta_time

TemporalOroboros commented:
@Cyberboss Conflixed

LemonInTheDark commented:
> 
> 
> > I'm going to blame mothblocks for this, even though it's my own fault.
> 
> WHAT

You know what you didn't do

LemonInTheDark commented:
I still have one or two unaddressed questions, but nothing major about behavior. Could do with a final testmerge before they're addressed and we merge

Cyberboss commented:
On all servers

Mothblocks commented:
look it hasn't exploded yet, fix the conflicts and i'll merge it if neither lemon or cyberboss have objections

LemonInTheDark commented:
I have one or two unaddressed reviews, but they're simple as hell. Should be good to go by tomorrow

LemonInTheDark commented:
Shoutout to [this commit ](https://github.com/tgstation/tgstation/commit/eeacf011867bd448e4119f7c26fcb7a0fad6eaf9)from 8 years ago for being fucking based to the core

TemporalOroboros committed:
Merge branch 'master' into life-delta-time

TemporalOroboros committed:
Fixes comments mentioning rates per tick

Cyberboss committed:
Merge branch 'master' into life-delta-time
